man1/psn.1
.TH psn 1 "2019-05-01" "PSN" "Process Snapper User Manual"
.SH NAME
psn \- Linux Process Snapper
.SH SYNOPSIS
.B psn [-h] [--list]
.br
.B psn [options]

.SH DESCRIPTION
Linux Process Snapper (pSnapper, psn) is a Linux `/proc` profiler that
works by sampling Linux task states and other metrics 
from `/proc/PID/task/TID` pseudofiles. pSnapper is a
.I passive sampling profiler
, it does not attach to your program to slow it down,
nor alter your program execution path or signal handling 
(like `strace` may inadvertently do).

As pSnapper is just a python script reading /proc files, it does not
require software installation, nor install any kernel modules. 
pSnapper does not even require root access in most cases. The exception 
is if you want to sample some “private” /proc files (like syscall,
and kernel stack) of processes running under other users.

.SH OPTIONS
.TP
.B \-h, --help 
show this help message and exit
.TP
.B  --list
list all available columns
.TP
.B \-d <seconds>
number of seconds to sample for
.TP
.B -p [pid], --pid [pid]
process id to sample (including all its threads), or
process name regex, or omit for system-wide sampling
.TP
.B -t [tid], --thread [tid]
thread/task id to sample (not implemented yet)
.TP
.B  -r, --recursive
also sample and report for descendant processes
.TP
.B -a, --all-states
display threads in all states, including idle ones
.TP
.B --sample-hz SAMPLE_HZ
sample rate in Hz
.TP
.B --ps-hz PS_HZ
sample rate of new processes in Hz
.TP
.B -o filename, --output-sample-db filename
path of sqlite3 database to persist samples to,
defaults to in-memory/transient
.TP
.B -i filename, --input-sample-db filename
path of sqlite3 database to read samples from instead
of actively sampling
.TP
.B  -s csv-columns, --select csv-columns
additional columns to report
.TP
.B  -g csv-columns, --group-by csv-columns
columns to aggregate by in reports
.TP
.B  -G csv-columns, --append-group-by csv-columns
 default + additional columns to aggregate by in reports

.P
.SH EXAMPLE
.TP 4
psn -h
get help
.TP
psn --list
show all possible reporting columns
.TP
psn
run pSnapper for 5 seconds, report task comm,state profile (default)
.TP
psn -d 10
same as above, but for 10 seconds
.TP
psn -p 1234
run pSnapper for PID 1234
.TP
psn -p 1234 -r
run pSnapper for PID 1234 and its children
.TP
psn -p 1234 -a
as above, but show all thread states (including idle, interruptible sleep)
.TP
psn -p "post|kwork"
run psnapper on processes  matching regex "post|kwork"
.TP
sudo psn -g comm,state,syscall
customize sampled /proc columns (syscall is "private", so need sudo if sampling all procs)
.TP
psn -g cmdline,state,syscall,wchan
choose exact sampled /proc columns
.TP
psn -G syscall,wchan
append sampled columns in addition to the default "comm,state"
.TP
psn -G kstack
show kernel stack backtraces in addition to the default columns

.SH KNOWN ISSUES
None currently
.SH SEE ALSO
sqlite3(1), proc(5), python(1)

.SH AUTHOR
Written by Tanel Poder \fIhttps://blog.tanelpoder.com/psnapper/\fR
.br
If you have suggestions for improvements in this tool, please send them along via the above address.

.SH COPYRIGHT
Copyright © 2019 Tanel Poder,  License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.  There is NO WARRANTY, to the extent permitted by law.



CHANGELOG.md


0.14
======================
* report file names that are accessed with I/O syscalls with arg0 as the file descriptor
  - example: `sudo psn -G syscall,filename`
  - works with read, write, pread, fsync, recvmsg, sendmsg etc, but not with batch io syscalls like io_submit(), select() that may submit multiple fds per call

* no need to install kernel-headers package anymore as pSnapper now has the unistd.h file bundled with the install
  - no more exceptions complaining about missing unistd_64.h file
  - pSnapper still tries to use the unistd.h file from a standard /usr/include location, but falls back to the bundled one if the file is missing. this should help with using pSnapper on other platforms too (different processor architectures, including 32bit vs 64bit versions of the same architecture have different syscall numbers

* pSnapper can now run on RHEL5 equivalents (2.6.18 kernel), however with separately installed python26 or later, as I haven't "downgraded" pSnapper's python code to work with python 2.4 (yet)
  - you could install python 2.6 or 2.7 manually in your own directory or use the EPEL package: (yum install epel-release ; yum install python26 )
  - you will also need to uncomment the 2nd line in psn script (use #!/usr/bin/env/python26 instead of python)
  - note that 2.6.18 kernel doesnt provide syscall,file name and kstack sampling (but wchan is available)



0.13
======================
* kernel stack summary reporting - new column `kstack`
* wider max column length (for kstack)
* add `--list` option to list all available columns
* replace digits from `comm` column by default to collapse different threads of the same thing into one. you can use `comm2` to see the unedited process comm.



bash_completion/psn.bash
#============================================================================
# Title       : psn.bash
# Description : bash_completion file for process snapper
# Author      : Bart Sjerps <bart@outrun.nl>
# Usage       : Copy this file to /etc/bash_completion.d/
# License     : GPLv3+
# ---------------------------------------------------------------------------

_psn() {
  local -a opts shortopts longopts
  local cur prev disks
  _get_comp_words_by_ref cur prev

  shortopts=(h d p t r a o i s g G)
  longopts+=(pid thread recursive all-states sample-hz ps-hz output-sample-db input-sample-db list)
  
  opts=$(printf "\x2d%s " "${shortopts[@]}")
  opts+=$(printf "\x2d\x2d%s " "${longopts[@]}")

  case ${prev} in
    psn)                   COMPREPLY=($(compgen -W "$opts" -- ${cur})) ;;
    -h)                    ;;
    -d)                    COMPREPLY=($(compgen -W "2 5 10" -- ${cur})) ;;
    -p|--pid)              COMPREPLY=($(compgen -W "$(ps -eo comm)" -- ${cur})) ;;
    -t)                    ;;
    --list)                ;;
    -r|--recursive)        COMPREPLY=($(compgen -W "$opts" -- ${cur})) ;;
    -a|--all-states)       COMPREPLY=($(compgen -W "$opts" -- ${cur})) ;;
       --sample-hz)        ;;
       --ps-hz)            ;;
    -o|--output-sample-db) COMPREPLY=($(compgen -f -- ${cur})) ;;
    -i|--input-sample-db)  COMPREPLY=($(compgen -f -- ${cur})) ;;
    -s|-g|-G)              COMPREPLY=($(compgen -W "$(psn --list | awk 'NF==2 {print $1}')" -- ${cur})) ;;
    *)                     COMPREPLY=($(compgen -W "$opts" -- ${cur})) ;;
  esac
  return 0
}
complete -F _psn psn
